hub.go

```diff
@@ -67,60 +70,98 @@ func (hub *Hub) LastEventID() EventID {
 }
 
 func (hub *Hub) stackTop() *layer {
+	hub.RLock()
 	stack := hub.stack
 	if stack == nil || len(*stack) == 0 {
+		hub.RUnlock()
 		return nil
 	}
-	return (*stack)[len(*stack)-1]
+	top := (*stack)[len(*stack)-1]
+	hub.RUnlock()
+
+	return top
 }
 
 // Clone returns a copy of the current Hub with top-most scope and client copied over.
 func (hub *Hub) Clone() *Hub {
-	return NewHub(hub.Client(), hub.Scope().Clone())
+	clone := NewHub(hub.Client(), hub.Scope().Clone())
+	return clone
 }
 
 // Scope returns top-level `Scope` of the current `Hub` or `nil` if no `Scope` is bound.
 func (hub *Hub) Scope() *Scope {
+	hub.RLock()
 	top := hub.stackTop()
+	hub.RUnlock()
+
 	if top == nil {
 		return nil
 	}
-	return top.scope
+
+	top.RLock()
+	scope := top.scope
+	top.RUnlock()
+
+	return scope
 }
 
 // Scope returns top-level `Client` of the current `Hub` or `nil` if no `Client` is bound.
 func (hub *Hub) Client() *Client {
+	hub.RLock()
 	top := hub.stackTop()
+	hub.RUnlock()
+
 	if top == nil {
 		return nil
 	}
-	return top.client
+
+	top.RLock()
+	client := top.client
+	top.RUnlock()
+
+	return client
 }
 
 // PushScope pushes a new scope for the current `Hub` and reuses previously bound `Client`.
 func (hub *Hub) PushScope() *Scope {
-	scope := hub.Scope().Clone()
+	hub.RLock()
+	client := hub.Client()
```

Comment:
Is the Go reader write lock happy with acquiring a read lock on a thread that already has one?
You'll get another read lock for `.Client()` and `.Scope()`.